### PR TITLE
Better dependency resolution blog post

### PR DESCRIPTION
**Pull request by greysteil**
download

### CODE REVIEW

**Comment by greysteil** [comment]
Kinda long, but I *think* that's OK?

**Comment by greysteil** [comment]
@sinjo - would love your thoughts too.

File: blog/_posts/2017-08-02-improving-dependency-resolution-in-bundler.md [hunk 1]
```diff
@@ -0,0 +1,235 @@
...
+A quick look at the Bundler repo surfaced [these][bundler-issue-1]
+[issues][bundler-issue-2] and gave me certainty the bug was somewhere in
+Bundler's resolution logic. Brilliantly, that resolution logic is extracted out
+into its own gem, [Molinillo][molinillo]. Time to get bug hunting...
+
+### The naive approach (and failure)
+
+To start out with, I thought I might be able to find a simple mistake in
+Molinillo, without really needing to get to know it. I created a
+[pull request][failed-pr] that got the tests passing and fixed my issue locally
```
Comment (Sinjo) [inline blog/_posts/2017-08-02-improving-dependency-resolution-in-bundler.md]: Missing full stop at end of sentence.

File: blog/_posts/2017-08-02-improving-dependency-resolution-in-bundler.md [hunk 2]
```diff
@@ -0,0 +1,235 @@
...
+Fundamentally, Molinillo resolves dependencies using a backtracking algorithm.
+If, like me, you have no idea what that is, here's the mental model that I used
+when thinking about it:
+
+- Take each "requirement" from a project's `Gemfile` and `Gemfile.lock` and put
+them in a big array. Each requirement will relate to a single dependency.
+- Dynamically build a "tree" from the requirements array as follows:
+  - Pick a requirement from the array (intelligently, but more on that later)
+  to become a node in the tree
+  - For each dependency version that satisfies the requirement, and all previous
```
Comment (Sinjo) [inline blog/_posts/2017-08-02-improving-dependency-resolution-in-bundler.md]: First time I read through, I spent a while wondering "what about shared transitive dependencies?" Just spotted the "and all previous requirements" bit on my second read. Might be worth calling it out slightly stronger here?

File: blog/_posts/2017-08-02-improving-dependency-resolution-in-bundler.md [hunk 3]
```diff
@@ -0,0 +1,235 @@
...
+apply the filtering we used during the decision process to actually remove the
+edges we know will lead us to the same confict as before
+
+Applying the above filtering process makes Molinillo just as snappy with the
+fixes previously discussed as it was before their application. Here's the
+[pull request][fix-pr] that makes it all happen.
+
+### Job done
+
+And with those changes, Molinillo, can resolve anything that's thrown at it,
```
Comment (Sinjo) [inline blog/_posts/2017-08-02-improving-dependency-resolution-in-bundler.md]: I think the comma after "Molinillo" can go.

**Pull request merged**